File: extract_data.js
// extract_data.js - Node script to extract all content from TS files into JSON
const fs = require('fs');
const path = require('path');

const BASE = path.join(__dirname, 'src', 'data');
const FILES = ['history.ts', 'culture.ts', 'philosophy.ts', 'science.ts', 'industry.ts', 'countries.ts', 'lifestyle.ts'];

function escapeJS(str) {
  return str.replace(/\\/g, '\\\\').replace(/'/g, "\\'").replace(/\n/g, '\\n');
}

function parseSections(text) {
  const sections = [];
  const re = /\{\s*id:\s*'([^']+)',\s*title:\s*'([^']+)'(?:,\s*content:\s*'([^']*)')?(?:,\s*subs:\s*\[([\s\S]*?)\])?\s*\}/g;
  let m;
  while ((m = re.exec(text)) !== null) {
    const sec = { id: m[1], title: m[2] };
    if (m[3] !== undefined) sec.content = m[3].replace(/\\'/g, "'");
    if (m[4] !== undefined) sec.subs = parseSections(m[4]);
    sections.push(sec);
  }
  return sections;
}

function extractCategory(filePath) {
  const text = fs.readFileSync(filePath, 'utf8');
  
  const catId = text.match(/id:\s*'([^']+)'/)?.[1] || '';
  const catTitle = text.match(/title:\s*'([^']+)'/)?.[1] || '';
  const catIcon = text.match(/icon:\s*'([^']+)'/)?.[1] || '';
  const catColor = text.match(/color:\s*'([^']+)'/)?.[1] || '';
  const catDesc = text.match(/description:\s*'([^']+)'/)?.[1] || '';
  
  // Find all topics
  const topics = [];
  const topicRegex = /\{\s*id:\s*'([^']+)',\s*title:\s*'([^']+)',\s*category:\s*'[^']+',\s*description:\s*'([^']+)',\s*icon:\s*'([^']+)',\s*sections:\s*\[([\s\S]*?)\]\s*\}/g;
  
  let m;
  while ((m = topicRegex.exec(text)) !== null) {
    const topic = {
      id: m[1],
      title: m[2],
      desc: m[3].replace(/\\'/g, "'"),
      icon: m[4],
      sections: parseSections(m[5])
    };
    topics.push(topic);
  }
  
  return { id: catId, title: catTitle, icon: catIcon, color: catColor, desc: catDesc, topics };
}

const categories = FILES.map(f => extractCategory(path.join(BASE, f)));

const outputPath = path.join(__dirname, 'docs', 'mikipedia-data.json');
fs.writeFileSync(outputPath, JSON.stringify(categories, null, 2), 'utf8');

console.log(`Extracted ${categories.length} categories, ${categories.reduce((s,c) => s + c.topics.length, 0)} topics`);
console.log(`Output: ${outputPath} (${fs.statSync(outputPath).size} bytes)`);

File: extract_data.py
#!/usr/bin/env python3
"""Extract all topic/section data from TS files into a single JSON for the web app."""
import re, json, os

BASE = '/Users/apple/Desktop/work/CSM/mikipedia/src/data'
FILES = ['history.ts', 'culture.ts', 'philosophy.ts', 'science.ts', 'industry.ts', 'countries.ts', 'lifestyle.ts']

def extract_categories():
    categories = []
    for fname in FILES:
        path = os.path.join(BASE, fname)
        with open(path, 'r') as f:
            text = f.read()
        
        # Extract category-level fields
        cat_id = re.search(r"id:\s*'([^']+)'", text).group(1)
        cat_title = re.search(r"title:\s*'([^']+)'", text).group(1)
        cat_icon = re.search(r"icon:\s*'([^']+)'", text).group(1)
        cat_color = re.search(r"color:\s*'([^']+)'", text).group(1)
        cat_desc = re.search(r"description:\s*'([^']+)'", text).group(1)
        
        # Extract topics
        topics = []
        # Find all topic blocks
        topic_blocks = re.findall(r'{\s*id:\s*\'([^\']+)\',\s*title:\s*\'([^\']+)\',\s*category:', text)
        topic_ids = [m[0] for m in topic_blocks]
        topic_titles = [m[1] for m in topic_blocks]
        
        # Find topic descriptions and icons
        topic_descs = re.findall(r"category:\s*'[^']+',\s*description:\s*'([^']+)'", text)
        topic_icons = re.findall(r"icon:\s*'([^']+)',\s*sections:", text)
        
        # Extract sections for each topic
        # Split by topic boundaries
        parts = re.split(r'{\s*id:\s*\'([^\']+)\',\s*title:\s*\'([^\']+)\',\s*category:', text)
        
        topic_idx = 0
        for i in range(1, len(parts), 2):
            tid = parts[i]
            ttitle = parts[i+1]
            block = parts[i+2] if i+2 < len(parts) else ''
            
            # Find the sections array
            sections_match = re.search(r'sections:\s*\[(.*)\]', block, re.DOTALL)
            if sections_match:
                sections_text = sections_match.group(1)
                sections = parse_sections(sections_text)
            else:
                sections = []
            
            desc = topic_descs[topic_idx] if topic_idx < len(topic_descs) else ''
            icon = topic_icons[topic_idx] if topic_idx < len(topic_icons) else ''
            
            topics.append({
                'id': tid,
                'title': ttitle,
                'desc': desc,
                'icon': icon,
                'sections': sections
            })
            topic_idx += 1
        
        categories.append({
            'id': cat_id,
            'title': cat_title,
            'icon': cat_icon,
            'color': cat_color,
            'desc': cat_desc,
            'topics': topics
        })
    
    return categories

def parse_sections(text):
    """Parse sections array text into structured list."""
    sections = []
    # Find each section block
    section_pattern = r'{\s*id:\s*\'([^\']+)\',\s*title:\s*\'([^\']+)\'(?:,\s*content:\s*\'((?:[^\'\\]|\\.)*)\')?(?:,\s*subs:\s*\[(.*?)\])?\s*}'
    
    for match in re.finditer(section_pattern, text, re.DOTALL):
        sid = match.group(1)
        stitle = match.group(2)
        content = match.group(3)
        subs_text = match.group(4)
        
        if content:
            # Unescape quotes
            content = content.replace("\\'", "'")
        
        subs = []
        if subs_text:
            subs = parse_sections(subs_text)
        
        section = {'id': sid, 'title': stitle}
        if content:
            section['content'] = content
        if subs:
            section['subs'] = subs
        
        sections.append(section)
    
    return sections

if __name__ == '__main__':
    cats = extract_categories()
    output_path = os.path.join(BASE, '..', '..', 'docs', 'mikipedia-data.json')
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(cats, f, ensure_ascii=False, indent=2)
    print(f"Extracted {len(cats)} categories with {sum(len(c['topics']) for c in cats)} topics to {output_path}")
    print(f"File size: {os.path.getsize(output_path)} bytes")
